### ISSUE TITLE
Accuracy levels after training don't agree with prediction metrics

### ISSUE THREAD

**Issue by nherger** [awslabs/sockeye#708]
Hello,
I'm trying to understand the difference between the validation accuracy score shown in the _metrics_ file in my sockeye model folder and the accuracy score I computed myself using sklearn after calling `sockeye.translate`.

The last line of my metrics file (written after model training was complete) look as follows:

> **accuracy-train=0.9997137355804897	accuracy-val=0.9394933114714763**	avg-sec-per-sent-val=0.013262660026550292	bleu-val=0.8434612769803594	chrf-val=0.9255211642727769	converged=True	decode-walltime-val=6.6313300132751465	diverged=False	epoch=421	gradient-norm=0.002509626388
> 760846	learning-rate=0.0002	length-ratio-mse=0.9971666666666668	perplexity-train=1.1135942407567943	perplexity-val=1.40864625280551	rouge1-val=0.5020857094907448	rouge2-val=0.10826666570746667	rougel-val=0.5014327228322503	time-elapsed=17690.11257648468	used-gpu-memory=8339

I then run `sockeye.translate` with the same source validation set (_val.source_) as given as input to `sockeye.train` with the `-vs` argument. The accuracy score obtained when comparing the output of the `sockeye.translate` function with the target validation set (_val.target_) is as follows:

> accuracy_score(stats.y_true, stats.y_pred)
> **0.8440152180275096**

How can there be such a difference between the accuracy-val from the metrics file (0.939) and the accuracy score I computed myself (0.844)?

Thanks,
Nadja

**Comment by fhieber** [comment]
Hi, thanks for your question!
My guess is that the difference in accuracy scores  is the effect of beam search at translation time and the fact that the model is never exposed to its own outputs during training time:
At training (sockeye.train), we 'teacher force' the model, i.e. we feed it the correct previous word at every time step on the target side. The accuracy computed here just measures how well the model is able to assign the maximum probability for the right next token, given the previous ground-truth token. At translation time (sockeye.translate), the model predicts the next token, given its previous prediction. Since that previous prediction might be wrong, the choice in the current time step might be suboptimal. As a result, the overall accuracy is lower.

**Comment by nherger** [comment]
I see. Thanks a lot for the thorough explanation!
